### PR TITLE
Split nwnx4 files between user and install dir

### PR DESCRIPTION
**Pull request by CromFr**
Allows NWNX4 files to be split into two directories:
- **Install directory** (e.g. `C:\Program Files\NWNX4\`)
    This directory contains all nwnx4 files from the nwnx4 release zip. Those files should not be modified by the user when configuring NWNX4 for their server. This is very similar to the NWN2 install directory.
- **User directory** (e.g. `Documents\NWNX4-myserver\`)
    This directory contains all configuration files, custom plugins, log files, and other files created by nwnx4. This is very similar to the NWN2 home directory (that is located in `Documents/Neverwinter Nights 2` by default).

This should help a lot for installing and updating nwnx4 while keeping track of files that have been added by the user.
A typical install would have a read-only NWNX4 install dir (placed by the admin or an installer inside Program Files), and a writable NWNX4 user dir placed in any handy location.
Multiple NWNX4 user dirs can share the same install dir. This is very practical if you need to run multiple nwn2 servers on the same machine.

The user directory is set to the current directory when launching the nwnx4 executable (controller or gui). If the user directory is empty, it is populated with the example config files and useful sub-folders.

Here are two methods for launching NWNX4 with a specific user directory:
1. Create a shortcut to launch NWNX4, edit its properties and change the "Start in" field to point to your user directory.
2. Create a .bat file to launch the nwnx4, containing:
    ```bat
    REM Moving to the user directory
    cd %USERPROFILE%\Documents\NWNX4-myserver
    REM Starting NWNX4
    C:\Program Files (x86)\nwnx4\NWNX4_Gui.exe
    ```

### Compatibility notes

Launching a nwnx4 executable with a shortcut or by double clicking the exe file, the user directory will be the same as the install directory, and everything should behave exactly like previous NWNX4 versions (the user and install directories are merged together).

Plugins only receives the nwnx4 user dir path when loaded (except for CPlugins), meaning they cannot reach files that are located in the nwnx4 install dir. This might cause some issues for some plugins, but I am not aware of any of such plugin.
However, if you encounter this issue, you can solve it by either copying the required files from the install dir into the user dir, or going back to the merged user+install dir setup.

If a plugin is present in both the install dir and user dir, the plugin from the user dir will be loaded. This can be useful if you need a custom xp_mysql plugin that is different from the official.

### CODE REVIEW

File: include/nwnx_cplugin.h [hunk 1]
```diff
@@ -11,8 +11,9 @@ struct NWNXCPlugin_InitInfo
 {
 	/// Path to this plugin DLL file
 	const char* dll_path;
-	/// Path to the NWNX4 base directory, where nwnx4_controller.exe is located.
-	const char* nwnx_path;
+    /// Path to the NWNX4 user directory, where config files are stored and log files should be
+    /// written.
+	const char* nwnx_user_path;
```
Comment (scottmunday84) [inline include/nwnx_cplugin.h:16]: We're telling them where the configurations are, but not where the base NWNX4 install is?

File: src/controller/controller.cpp [hunk 2]
```diff
@@ -265,31 +311,21 @@ void NWNXController::notifyServiceShutdown()
...
-	std::string nwnexe("\\nwn2server.exe");
-	auto pszHookDLLPath = "NWNX4_Hook.dll";
-
 	ZeroMemory(&si, sizeof(si));
 	ZeroMemory(&pi, sizeof(pi));
 	si.cb = sizeof(si);
 
-	auto exePath = nwninstalldir + nwnexe;
-	logger->Trace("Starting server executable %s in %s", exePath.c_str(), nwninstalldir.c_str());
+	auto exePath = m_nwn2InstallDir / "nwn2server.exe";
```
Comment (scottmunday84) [inline src/controller/controller.cpp:318]: That is cool! I didn't realize they overrode the / operator for paths like that.

File: src/controller/nwnx.ini [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # NWNX4 configuration file
-# These are the default values for NWNX4. 
+# These are the default values for NWNX4.
```
Comment (scottmunday84) [inline src/controller/nwnx.ini:2]: Looks to be a spacing fix. We should setup a PR to fix and cleanup the look of this file at a point. There's spelling and  grammatical errors that might help with readability.

**Comment by scottmunday84** [comment]
Overall, I like it, although one suggestion I would do is verify that the spacing is consistent. Some if conditions have no space, some have space. There's also a matter of tabbing blocks of code. I'm approving.

**Comment by scottmunday84** [comment]
Good job!

File: src/controller/controller.cpp (hunk 2 above)
Comment (CromFr) [reply src/controller/controller.cpp:311]: It's a bit disturbing the first time you see it, but it's really practical :)

File: src/controller/nwnx.ini (hunk 3 above)
Comment (CromFr) [reply src/controller/nwnx.ini:2]: We need to setup clang-format on the project, and have it automatically reformat the code (and check it's formatted with the CI)

However reformatting the code wil cause merge issues (or at least some overhead) if there's unmerged branches, so I'd better do it when the project isn't very active

**Review by scottmunday84 — APPROVED** [review]
Looks really good. Good job!

File: src/controller/controller.cpp [hunk 4]
```diff
@@ -160,16 +160,46 @@ Return Value:
...
+}
+
 NWNXController::NWNXController(SimpleIniConfig* config)
 {
     this->config = config;
 
     // Setup temporary directories.
     this->setupTempDirectories();
 
+	m_nwnx4UserDir = std::filesystem::current_path();
```
[the remark this answers is not in the document]
Comment (CromFr) [reply src/controller/controller.cpp:172]: I added a `-userdir <PATH>` argument for this purpose. It also creates and populates the userdir if it doesn't exist :)

The argument isn't available for NWNX4_Gui.exe though

**Pull request merged**